explorer-server/code/blocks.js

```diff
@@ -140,42 +139,64 @@ const findClosest = (haystack, needle) => (
   ))
 );
 
-const generatePaginationArray = (max, slots) => {
-  const validSteps = [1, 5, 10, 100, 200, 300]; // usable increments
-  const stepValues = validSteps.map(step => Math.ceil(max / step)); // realm of possibilities
+const generatePaginationArray = (currentPage, max, slots) => {
+  let increments;
+  let roundingExceptions;
 
-  // find closest
-  let bestValueIndex = stepValues.indexOf(findClosest(stepValues, slots));
-  if (stepValues[bestValueIndex] === 1) { bestValueIndex-- } // go to prev increment
+  if (slots === 3) {
+    increments = [1, 100, 500, 1000, 2000, 4000];
+    roundingExceptions = [0];
+  }
+  else if (slots === 7) {
+    increments = [1, 10, 50, 100, 500, 1000, 2000, 4000];
+    roundingExceptions = [0];
+  }
+  else {
+    increments = [1, 2, 10, 20, 50, 100, 500, 1000, 2000, 4000];
+    roundingExceptions = [0, 1];
+  }
 
-  const bestStep = validSteps[bestValueIndex]
-  const bestValue = stepValues[bestValueIndex]
-  const pageArray = [bestStep];
+  let pageArray = [];
 
-  // generate page array
-  for (i = 0; i < bestValue - 1; i++) {
-    pageArray.push(pageArray.slice(-1)[0] + bestStep);
-  }
+  for (i = 0; i < Math.floor(slots / 2); i++) {
+    const currentIncrement = increments[i];
 
-  // if spilling over: shift everything to the right
-  if (pageArray.slice(-1)[0] > max) {
-    pageArray.pop()
-    pageArray.unshift(pageArray[0] - validSteps[bestValueIndex - 1]);
-  }
+    if (!currentIncrement || (currentPage - currentIncrement <= 1)) {
+      break;
+    }
+
+    const value = currentPage - currentIncrement
+    const precision = String(value).length - 1
 
-  // recursive case: under (there are slots left)
-  if (bestValue < slots) {
-    return generatePaginationArray(pageArray[0], slots - bestValue + 1).concat(pageArray.slice(1));
+    if (!(i in roundingExceptions)) {
+      pageArray.push(parseFloat(value.toPrecision(precision)));
+    } else {
+      pageArray.push(value);
+    }
   }
 
-  // normal case: over (criss cross edges until equal)
-  if (bestValue > slots) {
-    for (i = 0; i < bestValue - slots; i++) {
-      if (i % 2) { pageArray.pop() } else { pageArray.shift() }
+  pageArray = pageArray.reverse();
+  if (currentPage != 1) { pageArray.push(currentPage) };
+
+  const remainingSlots = slots - pageArray.length;
+  for (i = 0; i < remainingSlots; i++) {
+    const currentIncrement = increments[i];
+    const value  = currentPage + currentIncrement;
+
+    if (!currentIncrement || (value > max)) {
+      break;
+    }
+
+    const precision = String(value).length - 1
+
+    if (!(i in roundingExceptions)) {
```

Comment:
In checks if `roundingExceptions[i]`, not any `i` in `roundingExceptions` exists. This is very weird behavior.

Imagine someone changed it to `roundingExceptions = [0, 2];`, then `2 in roundingExceptions` returns `false`, which will confuse everyone involved.